mtags/src/main/scala/scala/meta/internal/mtags/MtagsIndexer.scala

```diff
@@ -11,18 +11,21 @@ import scala.meta.internal.semanticdb.Scala._
 import scala.meta.internal.semanticdb.SymbolInformation.Kind
 import scala.meta.internal.{semanticdb => s}
 
-trait MtagsIndexer {
+trait GenericMtagsIndexer[T <: EnrichedTextDocument] {
   def language: Language
   def indexRoot(): Unit
   def input: Input.VirtualFile
-  def index(): s.TextDocument = {
+  protected def documentToResult(doc: s.TextDocument): T
+  def index(): T = {
```

Comment:
Would it make sense to just have overrides accessible as separate method? Do we need to box it together with TextDocument?